### PR TITLE
Pv2 926 pv2 937

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Tests/Non-Levy Learner - Basic Day/Two Non Levy-Learners One Finishes Late And Other Finishes Early PV2-199.feature [hunk 1]
```diff
@@ -1,12 +1,9 @@
 ﻿Feature: Two Non Levy-Learners One Finishes Late And Other Finishes Early PV2-199
-
-@EndToEnd
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Tests/Non-Levy Learner - Basic Day/Two Non Levy-Learners One Finishes Late And Other Finishes Early PV2-199.feature]: This seems to be an important attribute, why is it removed?

File: src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/MappingTest.cs [hunk 2]
```diff
@@ -122,8 +122,9 @@ public void SetUp()
                         PriceEpisodeIdentifier = "pe-1",
                         PriceEpisodeValues = new PriceEpisodeValues
                         {
-
+                            PriceEpisodeAimSeqNumber = 1,
                             EpisodeStartDate = DateTime.Today.AddMonths(-12),
+                            EpisodeEffectiveTNPStartDate = DateTime.Today.AddMonths(-12),
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.EarningEvents.Application.UnitTests/Mapping/MappingTest.cs]: Make these values different for the test to prove they are taken from relevant source

File: src/SFA.DAS.Payments.Model.Core/PriceEpisode.cs [hunk 3]
```diff
@@ -22,7 +22,9 @@ public class PriceEpisode
         /// </summary>
         public decimal? TotalNegotiatedPrice4 { get; set; }
         public decimal AgreedPrice { get; set; }
-        public DateTime StartDate { get; set; }
+        //public DateTime StartDate {  get; set; }
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.Model.Core/PriceEpisode.cs]: still needed? Can it be removed?

File: src/SFA.DAS.Payments.RequiredPayments.Application/Infrastructure/Configuration/RequiredPaymentsProfile.cs [hunk 4]
```diff
@@ -105,17 +103,47 @@ public RequiredPaymentsProfile()
 
             CreateMap<PayableEarningEvent, CalculatedRequiredLevyAmount>()
                 .ForMember(x => x.ContractType, opt => opt.UseValue(ContractType.Act1))
-                .ForMember(requiredPayment => requiredPayment.AccountId, opt => opt.MapFrom(p => p.AccountId))
-                ;
+                .ForMember(x => x.Priority, opt => opt.Ignore())
+                .ForMember(x => x.ApprenticeshipId, opt => opt.Ignore())
+                .ForMember(x => x.ApprenticeshipPriceEpisodeId, opt => opt.Ignore())
+                //.ForAllOtherMembers(opt => opt.Ignore())
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.RequiredPayments.Application/Infrastructure/Configuration/RequiredPaymentsProfile.cs]: can this now be removed?

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/PayableEarningEventMatcher.cs [hunk 5]
```diff
@@ -169,11 +169,13 @@ protected override bool Match(PayableEarningEvent expectedEvent, PayableEarningE
                 expectedEvent.LearningAim.FrameworkCode != actualEvent.LearningAim.FrameworkCode ||
                 expectedEvent.LearningAim.PathwayCode != actualEvent.LearningAim.PathwayCode ||
                 expectedEvent.LearningAim.ProgrammeType != actualEvent.LearningAim.ProgrammeType ||
-                expectedEvent.LearningAim.StandardCode != actualEvent.LearningAim.StandardCode ||
-                expectedEvent.Priority != actualEvent.Priority ||
-                expectedEvent.AccountId != actualEvent.AccountId ||
-                expectedEvent.CommitmentId != actualEvent.CommitmentId ||
-                expectedEvent.CommitmentVersion != actualEvent.CommitmentVersion)
+                expectedEvent.LearningAim.StandardCode != actualEvent.LearningAim.StandardCode 
+                //TODO: should every period be checked as these properties have now moved to earning period
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/PayableEarningEventMatcher.cs]: I have created a Technical Task - PV2-950 to reinstate the check

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentEventMatcher.cs [hunk 6]
```diff
@@ -157,6 +157,7 @@ protected override IList<ProviderPaymentEvent> GetExpectedEvents()
 
         protected override bool Match(ProviderPaymentEvent expected, ProviderPaymentEvent actual)
         {
+            var onProgTypes = new[] { TransactionType.Balancing, TransactionType.Completion, TransactionType.Learning };
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentEventMatcher.cs]: Moved

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentModelMatcher.cs [hunk 7]
```diff
@@ -82,6 +82,7 @@ protected override IList<PaymentModel> GetExpectedEvents()
 
         protected override bool Match(PaymentModel expected, PaymentModel actual)
         {
+            var onProgTypes = new[] { TransactionType.Balancing, TransactionType.Completion, TransactionType.Learning };
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentModelMatcher.cs]: Moved

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Helpers/ApprenticeshipHelper.cs [hunk 8]
```diff
@@ -14,13 +14,30 @@ namespace SFA.DAS.Payments.AcceptanceTests.EndToEnd.Helpers
 {
     public static class ApprenticeshipHelper
     {
+        public static async Task AddApprenticeship(ApprenticeshipModel apprenticeship, IPaymentsDataContext dataContext)
+        {
+            await dataContext.Apprenticeship.AddAsync(apprenticeship).ConfigureAwait(false);
+            await dataContext.SaveChangesAsync().ConfigureAwait(false);
+        }
+
+        public static async Task UpdateApprenticeship(long apprenticeshipId, ApprenticeshipStatus status, List<ApprenticeshipPriceEpisodeModel> priceEpisodes, IPaymentsDataContext dataContext)
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Helpers/ApprenticeshipHelper.cs]: If we find an existing apprenticeship we update it to the latest state. Good point on the status that was missed out. Now Changed

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Helpers/ApprenticeshipHelper.cs [hunk 9]
```diff
@@ -14,13 +14,30 @@ namespace SFA.DAS.Payments.AcceptanceTests.EndToEnd.Helpers
...
+            var apprenticeship = await dataContext.Apprenticeship
+                                     .Include(a => a.ApprenticeshipPriceEpisodes)
+                                     .FirstOrDefaultAsync(a => a.Id == apprenticeshipId) 
+                                 ?? throw new InvalidOperationException($"Apprenticeship not found: {apprenticeshipId}");
+            apprenticeship.ApprenticeshipPriceEpisodes.ForEach(priceEpisode => priceEpisode.Removed = true);
+            apprenticeship.ApprenticeshipPriceEpisodes.AddRange(priceEpisodes);
+            await dataContext.SaveChangesAsync().ConfigureAwait(false);
+        }
+
         public static async Task AddApprenticeships(Apprenticeship apprenticeshipSpec, List<Apprenticeship> testSessionApprenticeships, IPaymentsDataContext dataContext, TestSession testSession)
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Helpers/ApprenticeshipHelper.cs]: Now removed

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs [hunk 10]
```diff
@@ -176,16 +176,29 @@ protected async Task AddOrUpdateTestApprenticeships(List<Apprenticeship> apprent
...
+
+            foreach (var group in groupedApprenticeships)
             {
-                var foundApprenticeship = Apprenticeships.SingleOrDefault(x => x.Identifier == apprenticeshipSpec.Identifier);
-                if (foundApprenticeship == null)
+                var specApprenticeship = Apprenticeships.FirstOrDefault(a => a.Identifier == group.Key);
+                if (specApprenticeship == null)
                 {
-                    await ApprenticeshipHelper.AddApprenticeships(apprenticeshipSpec, Apprenticeships, DataContext, TestSession);
+                    //use last apprenticeship to make sure it picks up the most recent status
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: The apprenticeship spec table is already ordered by the start and end date grouping will not change the initial ordering.

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs [hunk 11]
```diff
@@ -176,16 +176,29 @@ protected async Task AddOrUpdateTestApprenticeships(List<Apprenticeship> apprent
...
+                    specApprenticeship = group.LastOrDefault() ?? throw new InvalidOperationException($"No apprenticeships in group. Identifier: {group.Key}");
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: True changed

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs [hunk 12]
```diff
@@ -176,16 +176,29 @@ protected async Task AddOrUpdateTestApprenticeships(List<Apprenticeship> apprent
...
+                    apprenticeship.ApprenticeshipPriceEpisodes = group.Select(ApprenticeshipHelper.CreateApprenticeshipPriceEpisode).ToList();
+                    await ApprenticeshipHelper.AddApprenticeship(apprenticeship, DataContext).ConfigureAwait(false);
+                    specApprenticeship.CommitmentId = apprenticeship.Id;
+                    Apprenticeships.Add(specApprenticeship);
                 }
                 else
                 {
-                    await ApprenticeshipHelper.UpdateApprenticeships(foundApprenticeship.CommitmentId, apprenticeshipSpec, DataContext);
+                    var priceEpisodes = group.Select(ApprenticeshipHelper.CreateApprenticeshipPriceEpisode).ToList();
+                    var status = group.LastOrDefault()?.Status?.ToApprenticeshipPaymentStatus() ??
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: Changed

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs [hunk 13]
```diff
@@ -372,16 +387,16 @@ private List<PriceEpisode> GeneratePriceEpisodes(Aim aim, IList<Earning> earning
...
             {
                 var currentPriceEpisode = priceEpisodesForAim[i];
-                var tnpStartDate = orderedPriceEpisodes
+                var tnpStartDate =  orderedPriceEpisodes
                     .First(x => x.PriceEpisodeValues.PriceEpisodeTotalTNPPrice ==
                                 currentPriceEpisode.PriceEpisodeValues.PriceEpisodeTotalTNPPrice)
-                    .PriceEpisodeValues.EpisodeStartDate;
-                currentPriceEpisode.PriceEpisodeValues.EpisodeEffectiveTNPStartDate = tnpStartDate;
+                    .PriceEpisodeValues.EpisodeEffectiveTNPStartDate;
+                //currentPriceEpisode.PriceEpisodeValues.EpisodeEffectiveTNPStartDate = tnpStartDate;
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: Restored to the previous code just changed the  EpisodeStartDate to EpisodeEffectiveTNPStartDate .

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Tests/Employer Change of Circumstances/Learner changes employer,16wk gap between commitments, provider is paid for the first 12wk of gap - PV2-621.feature [hunk 14]
```diff
@@ -4,41 +4,44 @@ As a provider,
 I want provider earnings and payments where learner changes employer and there is a 16 week gap between commitments - provider receives payment during the first 12 weeks of the gap as they amend the ACT code and employment status code correctly, to be paid the correct amount
 So that I am accurately paid my apprenticeship provision.
 	
-Scenario Outline:Learner changes employer,16wk gap between commitments, provider is paid for the first 12wk of gap - PV2-621
+Scenario Outline:Learner changes employer,16wk gap between commitments, provider is paid for the first 12wk of gap - PV2-621Scenario Outline: Learner changes employer and there is a gap between commitments - provider receives full payment for first 12weeks during the gap as they amend the ACT code and employment status code correctly, then no payment for remaining 4weeks, as learner can only get paid for up to 12weeks gap.
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Tests/Employer Change of Circumstances/Learner changes employer,16wk gap between commitments, provider is paid for the first 12wk of gap - PV2-621.feature]: Good spot changed

**Comment by samuel-dev-dba** [comment]
@ESFA-Pawan  Spec now updated

File: src/SFA.DAS.Payments.Model.Core/PriceEpisode.cs (hunk 3 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.Model.Core/PriceEpisode.cs]: Removed

File: src/SFA.DAS.Payments.RequiredPayments.Application/Infrastructure/Configuration/RequiredPaymentsProfile.cs (hunk 4 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.RequiredPayments.Application/Infrastructure/Configuration/RequiredPaymentsProfile.cs]: removed

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentEventMatcher.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentEventMatcher.cs]: Could have been made static, but no urgency.

[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentEventMatcher.cs]: We both missed it, but there is already a method for it. EnumHelper.IsOnProgType

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentModelMatcher.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (kuzinator) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/EventMatchers/ProviderPaymentModelMatcher.cs]: We both missed it, but there is already a method for it. EnumHelper.IsOnProgType

**Pull request merged**